circuits/src/zk_circuits/valid_wallet_update.rs

```diff
@@ -0,0 +1,1242 @@
+            compute_wallet_share_nullifier(old_public_commitment, old_wallet.blinder);
+
+        // Commit to the new private shares
+        let new_private_shares_commitment =
+            compute_wallet_share_commitment(new_wallet_private_shares.clone());
+
+        let witness = SizedWitness {
+            old_wallet_private_shares,
+            old_wallet_public_shares,
+            new_wallet_private_shares,
+            private_shares_opening: openings.remove(0),
+            public_shares_opening: openings.remove(0),
+        };
+        let statement = SizedStatement {
+            old_private_shares_nullifier: old_private_nullifier,
+            old_public_shares_nullifier: old_public_nullifier,
+            old_pk_root: old_wallet.keys.pk_root,
+            new_private_shares_commitment,
+            new_public_shares: new_wallet_public_shares,
+            merkle_root,
+            external_transfer,
+            timestamp: NEW_TIMESTAMP,
+        };
+
+        (witness, statement)
+    }
+
+    /// Return true if the circuit constraints are satisfied on a given
+    /// statement, witness pair
+    fn constraints_satisfied(statement: SizedStatement, witness: SizedWitness) -> bool {
+        // Build a constraint system
+        let pc_gens = PedersenGens::default();
+        let mut transcript = Transcript::new(b"test");
+        let mut prover = Prover::new(&pc_gens, &mut transcript);
+
+        // Allocate the witness and statement in the constraint system
+        let mut rng = OsRng {};
+        let statement_var = statement.commit_public(&mut prover).unwrap();
+        let (witness_var, _) = witness.commit_witness(&mut rng, &mut prover).unwrap();
+
+        // Apply the constraints
+        ValidWalletUpdate::circuit(statement_var, witness_var, &mut prover).unwrap();
+        prover.constraints_satisfied()
+    }
+
+    // ------------------------------
+    // | Order Placement Test Cases |
+    // ------------------------------
+
+    /// Tests a valid witness and statement for placing an order
+    #[test]
+    fn test_place_order__valid() {
+        let mut old_wallet = INITIAL_WALLET.clone();
+        let mut new_wallet = INITIAL_WALLET.clone();
+        new_wallet.orders[0].timestamp = NEW_TIMESTAMP;
+
+        // Remove an order from the initial wallet
+        old_wallet.orders[0] = Order::default();
+
+        assert!(constraints_satisfied_on_wallets(
+            old_wallet,
+            new_wallet,
+            ExternalTransfer::default()
+        ));
+    }
+
+    /// Tests order cancellation
+    #[test]
+    fn test_order_cancellation() {
+        let old_wallet = INITIAL_WALLET.clone();
+        let mut new_wallet = INITIAL_WALLET.clone();
+        new_wallet.orders[0] = Order::default();
+
+        assert!(constraints_satisfied_on_wallets(
+            old_wallet,
+            new_wallet,
+            ExternalTransfer::default()
+        ))
+    }
+
+    /// Tests placing an order without updating the timestamp
```

Comment:
this test feels a little strange to me... if i'm placing a new order, how can you tell if the timestamp has been "updated" or not? there's no old order to compare to.
unless the design here is that all "zero orders" come with a default timestamp that new orders must update from. if so:
1. is that only the case during testing?
2. if it's true in prod too, what is the default timestamp?

Reply:
No default timestamp, this is testing the case in which a prover lies about the public input timestamp that the contract will verify. I'll update this comment to be more clear, thanks for pointing out